spec/formatting.md

```diff
@@ -18,3 +18,236 @@ the local variable takes precedence.
 
 It is an error for a local variable definition to
 refer to a local variable that's defined after it in the message.
+
+## Error Handling
+
+During the formatting of a message,
+various errors may be encountered.
+These are divided into the following categories:
+
+- **Syntax errors** occur when the syntax representation of a message is not well-formed.
+
+  Example invalid messages resulting in a Syntax error:
+
+  ```
+  {Missing end brace
+  ```
+
+  ```
+  {Unknown {?placeholder}}
+  ```
+
+  ```
+  let $var = {(no message body)}
+  ```
+
+- **Data Model errors** occur when a message is invalid due to
+  violating one of the following semantic requirements on its structure:
+
+  - **Variant Key Mismatch errors** occur when the number of keys on a Variant
+    does not equal the number of Selectors.
+
+    Example invalid messages resulting in a Variant Key Mismatch error:
+
+    ```
+    match {$one}
+    when 1 2 {Too many}
+    when * {Otherwise}
+    ```
+
+    ```
+    match {$one} {$two}
+    when 1 2 {Two keys}
+    when * {Missing a key}
+    when * * {Otherwise}
+    ```
+
+  - **Missing Fallback Variant errors** occur when the message
+    does not include a Variant with only catch-all keys.
+
+    Example invalid messages resulting in a Missing Fallback Variant error:
+
+    ```
+    match {$one}
+    when 1 {Value is one}
+    when 2 {Value is two}
+    ```
+
+    ```
+    match {$one} {$two}
+    when 1 * {First is one}
+    when * 1 {Second is one}
+    ```
+
+- **Resolution errors** occur when the runtime value of a part of a message
+  cannot be determined.
+
+  - **Unresolved Variable errors** occur when a variable reference cannot be resolved.
+
+    For example, attempting to format either of the following messages
+    must result in an Unresolved Variable error if done within a context that
+    does not provide for the variable reference `$var` to be successfully resolved:
+
+    ```
+    {The value is {$var}.}
+    ```
+
+    ```
+    match {$var}
+    when 1 {The value is one.}
+    when * {The value is not one.}
+    ```
+
+  - **Unknown Function errors** occur when an Expression includes
```

Comment:
Again nitpick / editorial (if you want I can create a PR after this, or you do it):
I think that consistently marking things that come from the spec proper as "code" would help.
For example "Expression handler" vs "expression handler" made a difference in clarity.
Probably "`Expression` handler"  would help even more.

Similar for the error classes themselves, maybe quoted (or bolded, or links)
Compare
```
When an error occurs within a Selector, the selector must not match any VariantKey other than the catch-all * and a Resolution or Selector error is emitted.
```
vs
```
When an error occurs within a `Selector`, the selector must not match any `VariantKey` other than the catch-all * and a [Resolution error]() or [Selector error]() is emitted.
```



Note: global improvement

Reply:
We should discuss this together with #331 keywords. Very happy to follow any common style guide we come up with for this.